src/v/kafka/server/snc_quota_manager.cc

```diff
@@ -0,0 +1,696 @@
+    connection_throttle_until = now + res.request;
+
+    return res;
+}
+
+void snc_quota_manager::record_request_tp(
+  const size_t request_size, const clock::time_point now) noexcept {
+    _shard_quota.in.use(request_size, now);
+}
+
+void snc_quota_manager::record_response_tp(
+  const size_t request_size, const clock::time_point now) noexcept {
+    _shard_quota.eg.use(request_size, now);
+}
+
+ss::lowres_clock::duration
+snc_quota_manager::get_quota_balancer_node_period() const {
+    const auto v = _kafka_quota_balancer_node_period();
+    // zero period in config means do not run balancer
+    if (v == std::chrono::milliseconds::zero()) {
+        return std::chrono::duration_cast<std::chrono::milliseconds>(
+          ss::lowres_clock::duration::max() / 2);
+    }
+    return v;
+}
+
+void snc_quota_manager::update_node_quota_default() {
+    const ingress_egress_state<std::optional<quota_t>> new_node_quota_default
+      = calc_node_quota_default();
+    if (ss::this_shard_id() == quota_balancer_shard) {
+        // downstream updates:
+        // - shard effective quota (via _shard_quotas_update):
+        const auto calc_diff = [](
+                                 const std::optional<quota_t>& qold,
+                                 const std::optional<quota_t>& qnew) {
+            dispense_quota_amount diff;
+            if (qold && qnew) {
+                diff.amount = *qnew - *qold;
+                diff.amount_is = dispense_quota_amount::amount_t::delta;
+            } else if (qold || qnew) {
+                diff.amount = node_to_shard_quota(qnew);
+                diff.amount_is = dispense_quota_amount::amount_t::value;
+            }
+            return diff;
+        };
+        ingress_egress_state<dispense_quota_amount> shard_quotas_update{
+          .in = calc_diff(_node_quota_default.in, new_node_quota_default.in),
+          .eg = calc_diff(_node_quota_default.eg, new_node_quota_default.eg)};
+        ssx::spawn_with_gate(
+          _balancer_gate, [this, update = shard_quotas_update] {
+              return quota_balancer_update(update);
+          });
+    }
+    _node_quota_default = new_node_quota_default;
+    // - shard minimum quota:
+    update_shard_quota_minimum();
+}
+
+void snc_quota_manager::update_shard_quota_minimum() {
+    const auto f = [this](const std::optional<quota_t>& node_quota_default) {
+        const auto shard_quota_default = node_to_shard_quota(
+          node_quota_default);
+        return std::max<quota_t>(
+          _kafka_quota_balancer_min_shard_thoughput_ratio()
+            * shard_quota_default,
+          _kafka_quota_balancer_min_shard_thoughput_bps());
+    };
+    _shard_quota_minimum.in = f(_node_quota_default.in);
+    _shard_quota_minimum.eg = f(_node_quota_default.eg);
+    // downstream updates: none
+}
+
+void snc_quota_manager::maybe_arm_balancer_timer() {
+    if (_balancer_gate.is_closed()) {
```

Comment:
If the only call to this is within the `ssx::spawn_with_gate` then this is an impossible condition

Reply:
Another call to this is from `_kafka_quota_balancer_node_period.watch()`. But I see your point: gate can be checked only where needed, and that's fair.